statistics/sample_serial_test.go

```diff
@@ -0,0 +1,328 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package statistics
+
+import (
+	"math/rand"
+	"testing"
+	"time"
+
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/tidb/sessionctx/stmtctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/collate"
+	"github.com/pingcap/tidb/util/mock"
+	"github.com/pingcap/tidb/util/sqlexec"
+	"github.com/stretchr/testify/require"
+)
+
+type testSampleSuite struct {
+	count int
+	rs    sqlexec.RecordSet
+}
+
+func TestSample(t *testing.T) {
+	s := createTestSampleSuite()
+	t.Run("SubTestCollectColumnStats", SubTestCollectColumnStats(s))
+	t.Run("SubTestMergeSampleCollector", SubTestMergeSampleCollector(s))
+	t.Run("SubTestCollectorProtoConversion", SubTestCollectorProtoConversion(s))
+	t.Run("SubTestWeightedSampling", SubTestWeightedSampling(s))
+	t.Run("SubTestDistributedWeightedSampling", SubTestDistributedWeightedSampling(s))
+	t.Run("SubTestBuildStatsOnRowSample", SubTestBuildStatsOnRowSample(s))
+}
+
+func createTestSampleSuite() *testSampleSuite {
+	s := new(testSampleSuite)
+	s.count = 10000
+	rs := &recordSet{
+		data:      make([]types.Datum, s.count),
+		count:     s.count,
+		cursor:    0,
+		firstIsID: true,
+	}
+	rs.setFields(mysql.TypeLonglong, mysql.TypeLonglong)
+	start := 1000 // 1000 values is null
+	for i := start; i < rs.count; i++ {
+		rs.data[i].SetInt64(int64(i))
+	}
+	for i := start; i < rs.count; i += 3 {
+		rs.data[i].SetInt64(rs.data[i].GetInt64() + 1)
+	}
+	for i := start; i < rs.count; i += 5 {
+		rs.data[i].SetInt64(rs.data[i].GetInt64() + 2)
+	}
+	s.rs = rs
+	return s
+}
+
+func SubTestCollectColumnStats(s *testSampleSuite) func(*testing.T) {
+	return func(t *testing.T) {
+		sc := mock.NewContext().GetSessionVars().StmtCtx
```

Comment:
```suggestion
		t.Parallel()
		sc := mock.NewContext().GetSessionVars().StmtCtx
```

It seems that all test cases can be paralleled.

Reply:
after add `t.Parallel()` tests will fail, you can check, run mulit times.

Reply:
I found that the failure is caused by sharing and iterating `testSampleSuite.rs` in `SubTestCollectColumnStats`, `SubTestMergeSampleCollector`, and `SubTestMergeSampleCollector`.

I think we can keep these 3 tests in serial, but change the other 3 tests to be parallel, to make test cases as parallel as possible.

Reply:
thanks, updated.